pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +240,308 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+	blockEndAt := cycleEnd.Truncate(b.cfg.ConsumeInterval)
+	blockEnd := blockEndAt.UnixMilli()
+
+	var (
+		numBlocks     int
+		compactionDur time.Duration
+	)
+	defer func(t time.Time, startState partitionState) {
+		dur := time.Since(t)
+
+		// Don't propagate any errors derived from the canceled context.
+		if errors.Is(retErr, context.Canceled) {
+			retErr = nil
+		}
+
+		if retErr != nil {
+			level.Error(b.logger).Log("msg", "partition consumption failed", "partition", state.Partition, "dur", dur, "start_lag", startState.Lag, "end_lag", state.Lag, "err", retErr)
+			return
+		}
+
+		b.metrics.processPartitionDuration.WithLabelValues(fmt.Sprintf("%d", state.Partition)).Observe(dur.Seconds())
+		level.Info(b.logger).Log("msg", "done consuming", "partition", state.Partition, "dur", dur,
+			"start_lag", startState.Lag, "end_lag", state.Lag, "cycle_end", cycleEnd,
+			"last_block_end", time.UnixMilli(startState.LastBlockEnd), "curr_block_end", time.UnixMilli(state.LastBlockEnd),
+			"last_seen_till", startState.SeenTillOffset, "curr_seen_till", state.SeenTillOffset,
+			"num_blocks", numBlocks, "compact_and_upload_dur", compactionDur)
+	}(time.Now(), *state)
+
+	builder := NewTSDBBuilder(b.logger, b.cfg.DataDir, b.cfg.BlocksStorage, b.limits)
+	defer builder.Close()
+
+	level.Info(b.logger).Log("msg", "start consuming", "partition", state.Partition, "offset", state.Commit.At, "lag", state.Lag, "cycle_end", cycleEnd)
+
+	// TopicPartition to resume consuming on this iteration.
+	tp := map[string][]int32{b.cfg.Kafka.Topic: {state.Partition}}
+	b.kafkaClient.ResumeFetchPartitions(tp)
+	defer b.kafkaClient.PauseFetchPartitions(tp)
+
+	// We always rewind the partition's offset to the commit offset. This is so the cycle started exactly at the commit point,
+	// and not at what was (potentially over-) consumed previously.
+	b.seekPartition(state.Partition, state.Commit.At)
+
+	var (
+		firstRec  *kgo.Record
+		lastRec   *kgo.Record
+		commitRec *kgo.Record
+	)
+
+consumerLoop:
+	for remaining := state.Lag; remaining > 0; {
+		if err := context.Cause(ctx); err != nil {
+			return err
+		}
+		fetches := b.kafkaClient.PollFetches(ctx)
+		if fetches.IsClientClosed() {
+			level.Warn(b.logger).Log("msg", "client closed when fetching records")
+			return nil
+		}
+		fetches.EachError(func(_ string, _ int32, err error) {
+			if !errors.Is(err, context.Canceled) {
+				level.Error(b.logger).Log("msg", "failed to fetch records", "partition", state.Partition, "err", err)
+				b.metrics.fetchErrors.Inc()
+			}
+		})
+
+		numRecs := fetches.NumRecords()
+		remaining -= int64(numRecs)
+		b.metrics.fetchRecordsTotal.Add(float64(numRecs))
+
+		recIter := fetches.RecordIter()
+		for !recIter.Done() {
+			rec := recIter.Next()
+
+			if firstRec == nil {
+				firstRec = rec
+			}
+
+			level.Debug(b.logger).Log("msg", "consumed record", "offset", rec.Offset, "rec_ts", rec.Timestamp, "last_bmax", state.LastBlockEnd, "bmax", blockEnd)
```

Comment:
bmax isn't very informative if you haven't read the code. Can you rename it to `block_max_time` or something similar? Same with `last_bmax`